### PR TITLE
(STM32) ETZPC firewall controller updates

### PR DESCRIPTION
**Pull request by GseoC**
This P-R modifies the ETZPC driver so that the firewall configuration of the peripheral managed by ETZPC can be described in the device tree. This increases flexibility and allows per platform firewall configuration.

Moreover, add an ETZPC firewall bus probing mechanism that checks if the peripherals that are under the ETZPC and used in OP-TEE are secured before adding their drivers to the probe list. This is a protection against unsafe use of non secure peripherals in the secure world.

Some device tree cleaning as well.

### CODE REVIEW

File: core/arch/arm/dts/stm32mp13xc.dtsi [hunk 1]
```diff
@@ -11,6 +11,7 @@
 		interrupts = <GIC_SPI 80 IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&rcc CRYP1>;
 		resets = <&rcc CRYP1_R>;
+		access-controllers = <&etzpc STM32MP1_ETZPC_SAES_ID>;
```
Comment (etienne-lms) [inline core/arch/arm/dts/stm32mp13xc.dtsi:14]: `STM32MP1_ETZPC_CRYP_ID`

File: core/drivers/stm32_etzpc.c [hunk 2]
```diff
@@ -66,263 +61,400 @@
...
-	struct etzpc_hwcfg hwcfg = { };
+	const fdt32_t *cuint = NULL;
+	size_t i = 0;
+	int len = 0;
+
+	cuint = fdt_getprop(fdt, node, "default-access-conf", &len);
+	if (!cuint) {
+		DMSG("No ETZPC DECPROT configuration in DT");
+		return;
+	}
```
Comment (etienne-lms) [inline core/drivers/stm32_etzpc.c]: maybe `if (len & 1) panic("Invalid ETZPC configuration");`

File: core/drivers/stm32_etzpc.c [hunk 3]
```diff
@@ -66,263 +61,400 @@
...
+	struct io_pa_va base = { };
+	int len = 0;
+
+	fdt_fill_device_info(fdt, &etzpc_info, node);
+	if (etzpc_info.reg == DT_INFO_INVALID_REG ||
+	    etzpc_info.reg_size == DT_INFO_INVALID_REG_SIZE)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	base.pa = etzpc_info.reg;
+	etzpc_dev->pdata.name = fdt_get_name(fdt, node, &len);
```
Comment (etienne-lms) [inline core/drivers/stm32_etzpc.c]: should `strdup()` the name to not reference the DT once inits are completed.
Note to also free `etzpc_dev->pdata.name` in `stm32_etzpc_free()`.

**Review by etienne-lms — COMMENTED** [review]
For commit "dts: stm32: default disabled DMA at SoC level for stm32mp15 platforms":
Fix the commit message header line: s/disabled/disable/ and prefer an explicit message body: s/This node/DMA node in stm32mp15* SoC DTSI files/

Having an explicit message body (not requiring to read the header line for understand which part is changes) is a comment that applies to most of the commits in this series.

For commit "dts: stm32: disable ADC2 on stm32mp135f-dk":
in commit message: the change removes the ADC2 node configuration (incl. enable). The commit message should reflect that, something like: "Remove ADC2 configuration in stm32mp135-dk.dts since OP-TEE does not use the device".

For commit "dts: stm32: disable VREFBUF on stm32mp15x platforms"
Prefer an explicit message! "VREFBUF is currently not used **on stm32mp15*-dk* boards**, so disable it." Also fix the header line: s/stm32mp15x/stm32mp15 DKx/".

Few other comments, see below.

**Comment by etienne-lms** [comment]
I missed to post a comment on commit "drivers: stm32_etzpc: new driver to use firewall API":
There are many changes in _stm32_etzpc.c_ when it is moved from _core/drivers/_ to _core/drivers/firewall/_. Could it be possible to split to change in 2? 1 commit that moves the source file location (possibly with minor changes, that Git UIs/diff can show) and 1 that makes all the other changes (field/variable/function names etc...). If that's too tricky, I'm fine, tell me and I'll make my way to check and review the changes.

File: core/drivers/stm32_etzpc.c (hunk 2 above)
Comment (GseoC) [reply core/drivers/stm32_etzpc.c]: Reverted this change

File: core/drivers/stm32_etzpc.c [hunk 4]
```diff
@@ -66,263 +61,400 @@
...
-	pbase = fdt_reg_base_address(fdt, node);
-	if (pbase == DT_INFO_INVALID_REG)
-		panic();
+	if (!etzpc_dev)
+		panic("ETZPC probe failed");
+
+	res = init_etzpc_from_dt(etzpc_dev, fdt, node);
+	if (res) {
+		free(etzpc_dev->pdata.tzma_cfg);
+		free(etzpc_dev);
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_etzpc.c]: removed this API.

File: core/arch/arm/plat-stm32mp1/main.c [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #include <drivers/gic.h>
 #include <drivers/pinctrl.h>
 #include <drivers/stm32_etzpc.h>
+#include <drivers/firewall_device.h>
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/plat-stm32mp1/main.c:13]: useless include

**Comment by GseoC** [comment]
@etienne-lms Comments addressed in the new series but I didn't put the patches on top as I had to reorganise the commits. I added a fix for RNG as well and a patch for DH flavorlist.

@jneuhauser, let me know if that's ok for you. Peripheral should probe thanks to the simple-bus compatible but there will be no ETZPC config set. Maybe you do want to add an ETZPC configuration, we can discuss about that.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"dts: stm32: default disable DMA at SoC level for stm32mp15 platforms",
"dts: stm32: disable ADC2 on stm32mp135f-dk",
"dts: stm32: disable VREFBUF on stm32mp15-dkx platforms",
"dts: stm32: define ETZPC as an access controller for stm32mp15 platforms" (with a minor comment) and
"dts: stm32: define ETZPC as an access controller for stm32mp13 platforms".
"plat-stm32mp1: default enable CFG_DRIVERS_FIREWALL".

For commit `dts: stm32: use st,stm32mp15-i2c-non-secure compatible for the I2C4`, I need to chek the stm32_i2c driver.

For commit "dt-bindings: add platform specific ETZPC bindings": few review comments not addressed.

Comments for commit "dts: stm32: add the ETZPC configuration table for stm32mp1x boards".

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: stm32_etzpc: move the stm32_etzpc driver to the firewall folder" and 
"plat-stm32mp1: add CFG_STM32_ALLOW_UNSAFE_PROBE to probe unsafe peripherals".

File: core/arch/arm/dts/stm32mp135f-dk.dts [hunk 6]
```diff
@@ -68,6 +68,51 @@
 	};
 };
 
+&etzpc {
+	st,decprot =
+		<&etzpc DECPROT(STM32MP1_ETZPC_ADC1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
```
Comment (etienne-lms) [inline core/arch/arm/dts/stm32mp135f-dk.dts:73]: Remove `&etzpc` phandle references from `st,decprot` properties, here and in the other DTS files.

File: core/arch/arm/dts/stm32mp131.dtsi [hunk 7]
```diff
@@ -6,6 +6,7 @@
 
 #include <dt-bindings/clock/stm32mp13-clks.h>
 #include <dt-bindings/clock/stm32mp13-clksrc.h>
+#include <dt-bindings/firewall/stm32mp13-etzpc.h>
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/dts/stm32mp131.dtsi:9]: Discard this comment: this DTSI file indeed needs `stm32mp13-etzpc.h` (few commits later)

File: core/arch/arm/dts/stm32mp151.dtsi [hunk 8]
```diff
@@ -6,6 +6,7 @@
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/clock/stm32mp1-clks.h>
 #include <dt-bindings/reset/stm32mp1-resets.h>
+#include <dt-bindings/firewall/stm32mp15-etzpc.h>
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/dts/stm32mp151.dtsi:9]: Discard this comment: this DTSI file indeed needs stm32mp15-etzpc.h (few commits later)

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit 
"dts: stm32: use st,stm32mp15-i2c-non-secure compatible for the I2C4".

File: core/arch/arm/dts/stm32mp157c-ed1-scmi.dts [hunk 9]
```diff
@@ -11,6 +11,22 @@
...
+		<&etzpc DECPROT(STM32MP1_ETZPC_I2C6_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_RNG1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_HASH1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_CRYP1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRCTRL_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRPHYC_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_STGENC_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_BKPSRAM_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_IWDG1_ID, DECPROT_S_RW, DECPROT_LOCK)>;
+};
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/dts/stm32mp157c-ed1-scmi.dts:28]: Considering that, on SCMI boards, we secure the RCC, we may not have similar firewall configs. I'd prefer to keep it as-is.

File: core/arch/arm/dts/stm32mp157c-dk2-scmi.dts [hunk 10]
```diff
@@ -12,6 +12,22 @@
...
+		<&etzpc DECPROT(STM32MP1_ETZPC_I2C6_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_RNG1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_HASH1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_CRYP1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRCTRL_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRPHYC_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_STGENC_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_BKPSRAM_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_IWDG1_ID, DECPROT_S_RW, DECPROT_LOCK)>;
+};
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/dts/stm32mp157c-dk2-scmi.dts:29]: ditto

File: core/arch/arm/dts/stm32mp157a-dk1-scmi.dts [hunk 11]
```diff
@@ -12,6 +12,21 @@
...
+		<&etzpc DECPROT(STM32MP1_ETZPC_I2C4_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_I2C6_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_RNG1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_HASH1_ID, DECPROT_NS_RW, DECPROT_UNLOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRCTRL_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_DDRPHYC_ID, DECPROT_NS_R_S_W, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_STGENC_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_BKPSRAM_ID, DECPROT_S_RW, DECPROT_LOCK)>,
+		<&etzpc DECPROT(STM32MP1_ETZPC_IWDG1_ID, DECPROT_S_RW, DECPROT_LOCK)>;
+};
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/dts/stm32mp157a-dk1-scmi.dts:28]: ditto

File: core/arch/arm/dts/stm32mp135f-dk.dts (hunk 6 above)
Comment (GseoC) [reply core/arch/arm/dts/stm32mp135f-dk.dts:73]: Must be a mix on my side, sorry and good spot.

**Comment by GseoC** [comment]
Given the comments were straight forward, I didn't bother with additional review patches. I addressed and squashed them.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"dt-bindings: add platform specific ETZPC bindings",
"dts: stm32: add the ETZPC configuration table for stm32mp1x boards",
"drivers: stm32_etzpc: update driver to set ETZPC configuration from DT",
"drivers: stm32_etzpc: new driver to use firewall API" and 
"drivers: stm32_rng: embed ETZPC functions when CFG_STM32_ETZPC is set".

Regarding "drivers: stm32_etzpc: update driver to set ETZPC configuration from DT", now that the peripheral secure hardening configuration is defined by the DT, we will have to remove call to `stm32mp_register_[non_]secure_periph_iomem()` calls in platform and driver since they are not really useful.

**Comment by jneuhauser** [comment]
Hello and sorry for the long delay...

Unfortunately, I don't have time for an indepth review at the moment.

The changes related to our boards and the other boards are ok, so `Acked-by: Johann Neuhauser <jneuhauser@dh-electronics.com>` for commit "dts: stm32: add the ETZPC configuration table for stm32mp1x boards".

I wonder if it makes more sense to define the firewall configuration in the files `stm32mp15xx-dhcor-som.dtsi` and `stm32mp15xx-dhcom-som.dtsi`, which represent the "System on Module" soldered or attached to the baseboard?

Regards, Johann

**Comment by GseoC** [comment]
@jneuhauser 's tag applied. I also addressed your comment and placed ETZPC config in the dhcom/dhcor dtsi files.

@etienne-lms , when rebasing, I reverted the `326382a059a8 ("drivers: stm32_rng: MP15 RNG is non-secure when PRNG is enable")` patch and squashed it in the last. I kept your tag, let me know if that's ok

**Comment by etienne-lms** [comment]
Looks all good to me. All review tags are applied.

**Pull request merged**